File: src/test/java/pages/purchasePages/PurchasePricePage.java
package pages.purchasePages;

import com.microsoft.playwright.FrameLocator;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.LoadState;
import enums.DistributorInfo;
import org.junit.Assert;
import pages.commonPages.BasePage;

public class PurchasePricePage extends BasePage {

    Locator pageTitle = page.locator("#PageTitle");
    Locator newRecord = page.locator(".glyphicon.glyphicon-plus");
    FrameLocator purchasePriceFrame = getFrameByDialogTitle("Satınalma Fiyatı Oluştur");
    FrameLocator productDefFrame = getFrameByDialogTitle("Ürün Tanımlama");
    Locator priceFrameName = page.locator("#SeturModalWin_wnd_title");
    FrameLocator undefinedDistributorFirm = getFrameByDialogTitle("Firma Tanımlama");


    /**
     * Satın Alma Fiyatları sayfasını doğrular.
     */
    public void verifyPurchasePricePage() {
        verifyTextElementUseTrim("Satınalma Fiyatları",pageTitle);
    }

    /**
     * Satınalma Fiyatları yeni kayıt açar.
     */
    public void newRecordPurchasePrice() {
        clickElement(newRecord);
    }

    /**
     * Satınalma Fiyatı Oluştur frame doğrular
     */
    public void verifyCreatePurchasePriceFrame(){
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        verifyTextElement("Satınalma Fiyatı Oluştur", priceFrameName);
    }

    /**
     * Ürün Tanımlama frame açar.
     */
    public void openProductDescriptionFrame() {
        purchasePriceFrame.locator("#ProductIdButtonId").click();
    }

    /**
     * Ürün Tanımlama frame doğrular.
     */
    public void verifyProductDescFrame(){
        Locator productFrameName = page.locator("span.k-window-title", new Page.LocatorOptions()
                .setHasText("Ürün Tanımlama"));
        verifyTextElement("Ürün Tanımlama", productFrameName);
    }

    /**
     * Ürün Kodu alanını doldurur.
     */
    public void fillProductCode(String productCode) {
        setKendoNumericTextBoxValue(productDefFrame, "#FilterProductId", productCode);
    }

    /**
     * Girilen ürün kodunu sorgular.
     */
    public void searchProduct(){
        productDefFrame.locator("#FilterButtonId").click();
    }

    /**
     * Sorgu sonucu gelen ürünü seçer. Ürün Tanımlama frame kapanır.
     */
    public void selectProduct(){
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        productDefFrame.locator("td[data-field-name='ProductId'] input[type='button']").click();
    }

    /**
     *  Satınalma Fiyatı Oluştur frame başlangıç tarihini doldurur.
     */
    public void fillStartDate() {
        String randomDate = generateRandomDate();
        purchasePriceFrame.locator("#StartDate").fill(randomDate);
    }

    /**
     * Satınalma Fiyatı doldurur.
     */
    public void fillPurchasePrice(){
        int number = generateRandomNumber();
        setKendoNumericTextBoxValue(purchasePriceFrame,"#Amount", String.valueOf(number));
    }

    /**
     * Tutar değerini ilerde kullanmak için saklar.
     */
    public void getValueOfAmount(){
        String newAmount = purchasePriceFrame.locator("#Amount").getAttribute("aria-valuenow");
        addString("newAmount",newAmount);
    }

    /**
     * Fiyat Türü seçer.
     */
    public void selectPriceType(){
        purchasePriceFrame.locator("span.k-select > span.k-icon.k-i-arrow-s")
                .nth(1).click();
        purchasePriceFrame.locator("#PriceTypeId_listbox li").nth(1).click();
    }

    /**
     * Satınalma Fiyatı Oluştur işlemini kaydeder.
     */
    public void saveCreatePurchasePrice(){
        purchasePriceFrame.locator("#Save").click();
    }

    /**
     * Ana sayfada Ürün Tanımlam frame açar.
     */
    public void openProductDescFrame() {
        page.locator("#FilterProductIdButtonId").click();
    }

    /**
     * Ana sayfada ürün sorgular.
     */
    public void searchProductInMainPage(){
        page.locator("#FilterButtonId").click();
    }

    /**
     * Satın Alma Fiyatını karşılaştırır.
     */
    public void verifyPurchasePriceAmount() {
        String amount = getString("newAmount"); // örn: "9091"
        Locator mainPageAmount = page.locator("td[data-field-name='Amount']").nth(0);

        String actualText = mainPageAmount.textContent().replace(".", ""); // "9.091,000000" → "9091,000000"
        Assert.assertEquals(amount + ",000000", actualText);
    }

    /**
     * Tamınsız ürün seçer.
     */
    public void selectUndefinedProduct() {
        purchasePriceFrame.locator("#no_DefUndefProduct").click();
    }

    /**
     * Tanımsız ürün Firma Tanımı frame açar.
     */
    public void openCompanyIdentification(){
        purchasePriceFrame.locator("#UndefinedDistributorFirmIdButtonId").click();
    }

    /**
     * Kategori seçimine göre firma kodu doldurur.
     * @param distributorInfo DistributorInfo enum klasından istenilen firma kodu
     */
    public void fillCompanyCode(DistributorInfo distributorInfo){
        undefinedDistributorFirm.locator("#FilterFirmCode").fill(distributorInfo.getFirmCode());
    }

    /**
     * Firma sorgular.
     */
    public void searchCompany(){
        undefinedDistributorFirm.locator("#FilterButtonId").click();
    }

    /**
     * Sorgulanan firmayı seçer.
     */
    public void selectCompany(){
        undefinedDistributorFirm.locator("input[type='button'][name*='FirmGrid']").click();
    }

    /**
     * Tanımsız ürün için üretici firma Firma tanımlama frame açar.
     */
    public void openManufacturerCompany(){
        purchasePriceFrame.locator("#UndefinedProducerFirmIdButtonId").click();
    }
    /**
     * Kategori seçimine göre firma kodu doldurur.
     * @param distributorInfo DistributorInfo enum sınıfında istenilen firma kodu
     */
    public void fillManufacturerCompany(DistributorInfo distributorInfo){
        undefinedDistributorFirm.locator("#FilterFirmCode").fill(distributorInfo.getFirmCode());
    }

    /**
     * Barkod numarası üretir.
     */
    public void fillUnidentifiedProductBarcode() {
        String randomNumber = generateBarcodeNumber();
        purchasePriceFrame.locator("#UndefinedBarcode").fill(randomNumber);
    }

    /**
     * KDV tutarı doldurur.
     */
    public void setVatAmount(){
        setKendoNumericTextBoxValue(purchasePriceFrame,"#VatAmount","7");
    }

    /**
     * Ana sayfada firma tanımlama frame açar.
     */
    public void openCompanyIdentificationFrame() {
        page.locator("#FilterDistributorFirmIdButtonId").click();
    }
}

File: RetailTRUI.Tests/Pages/Purchasing/PurchasePricePage.cs
using RetailTRUI.Tests.Enums;
using RetailTRUI.Tests.Pages.Common;

namespace RetailTRUI.Tests.Pages.Purchasing;

/// <summary>
/// Purchase Price page object for creating and managing purchase prices
/// Supports both defined and undefined products
/// </summary>
public class PurchasePricePage : BasePage
{
    private ILocator PageTitle => Page.Locator("#PageTitle");
    private ILocator NewRecord => Page.Locator(".glyphicon.glyphicon-plus");
    private IFrameLocator PurchasePriceFrame => GetFrameByDialogTitle("Satınalma Fiyatı Oluştur");
    private IFrameLocator ProductDefFrame => GetFrameByDialogTitle("Ürün Tanımlama");
    private ILocator PriceFrameName => Page.Locator("#SeturModalWin_wnd_title");
    private IFrameLocator UndefinedDistributorFirm => GetFrameByDialogTitle("Firma Tanımlama");

    public async Task VerifyPurchasePricePageAsync()
    {
        var actualText = await PageTitle.TextContentAsync();
        actualText?.Trim().Should().Be("Satınalma Fiyatları");
    }

    public async Task NewRecordPurchasePriceAsync()
    {
        await NewRecord.ClickAsync();
        await VerifyCreatePurchasePriceFrameAsync();
    }

    public async Task VerifyCreatePurchasePriceFrameAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        var actualText = await PriceFrameName.TextContentAsync();
        actualText.Should().Be("Satınalma Fiyatı Oluştur");
    }

    public async Task CreatePurchasePriceForDefinedProductAsync()
    {
        await OpenProductDescriptionFrameAsync();
    }

    private async Task OpenProductDescriptionFrameAsync()
    {
        await PurchasePriceFrame.Locator("#ProductIdButtonId").ClickAsync();
    }

    public async Task SelectDefinedProductAsync(string productCode)
    {
        await VerifyProductDescFrameAsync();
        await FillProductCodeAsync(productCode);
        await SearchProductAsync();
        await SelectProductAsync();
    }

    private async Task VerifyProductDescFrameAsync()
    {
        var productFrameName = Page.Locator("span.k-window-title")
            .Filter(new LocatorFilterOptions { HasText = "Ürün Tanımlama" });
        var actualText = await productFrameName.TextContentAsync();
        actualText.Should().Be("Ürün Tanımlama");
    }

    private async Task FillProductCodeAsync(string productCode)
    {
        var input = ProductDefFrame.Locator("#FilterProductId");
        await input.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });
        
        var handle = await input.ElementHandleAsync();
        if (handle != null)
        {
            await Page.EvaluateAsync(@"
                (element, val) => {
                    const widget = $(element).data('kendoNumericTextBox');
                    if (widget) {
                        widget.value(val);
                        widget.trigger('change');
                    }
                }", new object[] { handle, productCode });
        }
    }

    private async Task SearchProductAsync()
    {
        await ProductDefFrame.Locator("#FilterButtonId").ClickAsync();
    }

    private async Task SelectProductAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        await ProductDefFrame.Locator("td[data-field-name='ProductId'] input[type='button']").ClickAsync();
    }

    public async Task FillPurchasePriceForDefinedProductAsync()
    {
        await FillStartDateAsync();
        await FillPurchasePriceAsync();
        await GetValueOfAmountAsync();
        await SelectPriceTypeAsync();
        await SaveCreatePurchasePriceAsync();
    }

    private async Task FillStartDateAsync()
    {
        var randomDate = GenerateRandomDate();
        await PurchasePriceFrame.Locator("#StartDate").FillAsync(randomDate);
    }

    private async Task FillPurchasePriceAsync()
    {
        var number = GenerateRandomNumber();
        
        var input = PurchasePriceFrame.Locator("#Amount");
        await input.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });
        
        var handle = await input.ElementHandleAsync();
        if (handle != null)
        {
            await Page.EvaluateAsync(@"
                (element, val) => {
                    const widget = $(element).data('kendoNumericTextBox');
                    if (widget) {
                        widget.value(val);
                        widget.trigger('change');
                    }
                }", new object[] { handle, number });
        }
    }

    private async Task GetValueOfAmountAsync()
    {
        var newAmount = await PurchasePriceFrame.Locator("#Amount").GetAttributeAsync("aria-valuenow");
        AddString("newAmount", newAmount!);
    }

    private async Task SelectPriceTypeAsync()
    {
        await PurchasePriceFrame.Locator("span.k-select > span.k-icon.k-i-arrow-s").Nth(1).ClickAsync();
        await PurchasePriceFrame.Locator("#PriceTypeId_listbox li").Nth(1).ClickAsync();
    }

    private async Task SaveCreatePurchasePriceAsync()
    {
        await PurchasePriceFrame.Locator("#Save").ClickAsync();
    }

    public async Task SearchDefinedProductAndVerifyAmountAsync(string productCode)
    {
        await OpenProductDescFrameAsync();
        await FillProductCodeAsync(productCode);
        await SearchProductAsync();
        await SelectProductAsync();
        await SearchProductInMainPageAsync();
        await VerifyPurchasePriceAmountAsync();
    }

    private async Task OpenProductDescFrameAsync()
    {
        await Page.Locator("#FilterProductIdButtonId").ClickAsync();
    }

    private async Task SearchProductInMainPageAsync()
    {
        await Page.Locator("#FilterButtonId").ClickAsync();
    }

    private async Task VerifyPurchasePriceAmountAsync()
    {
        var amount = GetString("newAmount");
        var mainPageAmount = Page.Locator("td[data-field-name='Amount']").Nth(0);

        var actualText = await mainPageAmount.TextContentAsync();
        var cleanedText = actualText?.Replace(".", ""); // "9.091,000000" → "9091,000000"
        cleanedText.Should().Be($"{amount},000000");
    }

    public async Task SelectPurchasePriceForUndefinedProductAsync()
    {
        await VerifyCreatePurchasePriceFrameAsync();
        await SelectUndefinedProductAsync();
    }

    private async Task SelectUndefinedProductAsync()
    {
        await PurchasePriceFrame.Locator("#no_DefUndefProduct").ClickAsync();
    }

    public async Task SelectDistributorCompanyAsync()
    {
        await OpenCompanyIdentificationAsync();
        await FillCompanyCodeAsync(DistributorInfo.TUTUN_URUNLERI);
        await SearchCompanyAsync();
        await SelectCompanyAsync();
    }

    private async Task OpenCompanyIdentificationAsync()
    {
        await PurchasePriceFrame.Locator("#UndefinedDistributorFirmIdButtonId").ClickAsync();
    }

    private async Task FillCompanyCodeAsync(DistributorInfo distributorInfo)
    {
        var firmCode = distributorInfo.GetFirmCode();
        await UndefinedDistributorFirm.Locator("#FilterFirmCode").FillAsync(firmCode);
    }

    private async Task SearchCompanyAsync()
    {
        await UndefinedDistributorFirm.Locator("#FilterButtonId").ClickAsync();
    }

    private async Task SelectCompanyAsync()
    {
        await UndefinedDistributorFirm.Locator("input[type='button'][name*='FirmGrid']").ClickAsync();
    }

    public async Task SelectUndefinedProductManufacturerCompanyAsync()
    {
        await OpenManufacturerCompanyAsync();
        await FillManufacturerCompanyAsync(DistributorInfo.TUTUN_URUNLERI);
        await SearchCompanyAsync();
        await SelectCompanyAsync();
    }

    private async Task OpenManufacturerCompanyAsync()
    {
        await PurchasePriceFrame.Locator("#UndefinedProducerFirmIdButtonId").ClickAsync();
    }

    private async Task FillManufacturerCompanyAsync(DistributorInfo distributorInfo)
    {
        var firmCode = distributorInfo.GetFirmCode();
        await UndefinedDistributorFirm.Locator("#FilterFirmCode").FillAsync(firmCode);
    }

    public async Task FillPurchasePriceForUndefinedProductAsync()
    {
        await FillUnidentifiedProductBarcodeAsync();
        await FillStartDateAsync();
        await FillPurchasePriceAsync();
        await GetValueOfAmountAsync();
        await SetVatAmountAsync();
        await SelectPriceTypeAsync();
        await SaveCreatePurchasePriceAsync();
    }

    private async Task FillUnidentifiedProductBarcodeAsync()
    {
        var randomNumber = GenerateBarcodeNumber();
        await PurchasePriceFrame.Locator("#UndefinedBarcode").FillAsync(randomNumber);
    }

    private async Task SetVatAmountAsync()
    {
        var input = PurchasePriceFrame.Locator("#VatAmount");
        await input.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });
        
        var handle = await input.ElementHandleAsync();
        if (handle != null)
        {
            await Page.EvaluateAsync(@"
                (element, val) => {
                    const widget = $(element).data('kendoNumericTextBox');
                    if (widget) {
                        widget.value(val);
                        widget.trigger('change');
                    }
                }", new object[] { handle, "7" });
        }
    }

    public async Task SearchUndefinedProductAndVerifyAmountAsync()
    {
        await OpenCompanyIdentificationFrameAsync();
        await FillCompanyCodeAsync(DistributorInfo.TUTUN_URUNLERI);
        await SearchCompanyAsync();
        await SelectCompanyAsync();
        await SearchProductInMainPageAsync();
        await VerifyPurchasePriceAmountAsync();
    }

    private async Task OpenCompanyIdentificationFrameAsync()
    {
        await Page.Locator("#FilterDistributorFirmIdButtonId").ClickAsync();
    }
}
